pkg/controller/networkconfig/cluster.go

```diff
@@ -0,0 +1,91 @@
+package networkconfig
+
+import (
+	"context"
+	"log"
+	"reflect"
+
+	configv1 "github.com/openshift/api/config/v1"
+	networkoperatorv1 "github.com/openshift/cluster-network-operator/pkg/apis/networkoperator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/apply"
+	"github.com/openshift/cluster-network-operator/pkg/names"
+	"github.com/openshift/cluster-network-operator/pkg/network"
+	k8sutil "github.com/openshift/cluster-network-operator/pkg/util/k8s"
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// MergeClusterConfig merges in the existing cluster config in to the
+// network operator config, overwriting any changes to the managed fields.
+func (r *ReconcileNetworkConfig) MergeClusterConfig(ctx context.Context, cfg *networkoperatorv1.NetworkConfig) error {
+	// fetch the cluster config
+	clusterConf := &configv1.Network{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: names.CLUSTER_CONFIG}, clusterConf)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// Validate config
+	// If invalid, do nothing - we should proceed
+	if err := network.ValidateClusterConfig(clusterConf.Spec); err != nil {
+		log.Printf("WARNING: ignoring cluster.openshift.io/v1/Network object - failed validation: %v", err)
+		return nil
+	}
+
+	existing := cfg.DeepCopy()
+
+	// Merge the cluster config in to the downstream CRD
+	network.MergeClusterConfig(&cfg.Spec, clusterConf.Spec)
+	if reflect.DeepEqual(cfg.Spec, existing.Spec) {
+		return nil
+	}
+
+	// If there are changes to the "downstream" networkconfig, commit it back
+	// to the apiserver
+	log.Println("WARNING: NetworkConfig.networkoperator.openshift.io has fields being overwritten by Network.cluster.openshift.io configuration")
+	cfg.TypeMeta = metav1.TypeMeta{APIVersion: "networkoperator.openshift.io/v1", Kind: "NetworkConfig"}
```

Comment:
why is this needed?

Reply:
maybe this is one of those things where you're mutating an object you got from a cache rather than DeepCopying it and mutating the copy?

Reply:
Probably not that, but something similar. The client always copies in to a receiving object, and all the types should be registered with the schema. Just confusing.